### ISSUE TITLE
Painless script don't have access to a _index variable and groovy is depreciated in 5.4

### ISSUE THREAD

**Issue by kodefear** [elastic/elasticsearch#24820]
**Elasticsearch version**: 5.4

**Plugins installed**: none

**JVM version**: 1.8.0_131

**OS version**:  Windows 10

Description of the problem including expected versus actual behavior:

According to the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-advanced-scripting.html), Text features, such as term or document frequency for a specific term can be accessed in scripts with the _index variable. When attempting this with a Painless script, I get a Variable [_index] is not defined. error.

**Steps to reproduce:**

Execute a query with a painless script that attempts to access _index, something like this:

```
GET /ap_dataset/document/_search
{
  "query": {
    "function_score": {
      "query": {
        "match": {
          "text": "cow"
        }
      },
      "boost_mode": "replace",
      "functions": [
        {
          "script_score": {
            "script": {
               "lang": "painless",
               "inline": "_index['text']['cow'].tf()"
            }
          }
        }
      ]
    }
  }
}
```

**Provide logs (if relevant):**

```
{
  "error": {
    "root_cause": [
      {
        "type": "script_exception",
        "reason": "compile error",
        "script_stack": [
          "_index['text']['cow'].tf( ...",
          "^---- HERE"
        ],
        "script": "_index['text']['cow'].tf()",
        "lang": "painless"
      }
    ],
    "type": "search_phase_execution_exception",
    "reason": "all shards failed",
    "phase": "query",
    "grouped": true,
    "failed_shards": [
      {
        "shard": 0,
        "index": "ap_dataset",
        "node": "zhCmVrZgRC-YQU5BOaz49w",
        "reason": {
          "type": "query_shard_exception",
          "reason": "script_score: the script could not be loaded",
          "index_uuid": "-TpUstdDQVSZnLwvLKWkPA",
          "index": "ap_dataset",
          "caused_by": {
            "type": "script_exception",
            "reason": "compile error",
            "script_stack": [
              "_index['text']['cow'].tf( ...",
              "^---- HERE"
            ],
            "script": "_index['text']['cow'].tf()",
            "lang": "painless",
            "caused_by": {
              "type": "illegal_argument_exception",
              "reason": "Variable [_index] is not defined."
            }
          }
        }
      }
    ]
  },
  "status": 400
} 
```
I tried using the groovy language, but it says 
**#! Deprecation: [groovy] scripts are deprecated, use [painless] scripts instead**

**Comment by rjernst** [comment]
Accessing the term dictionary and statistics has long been a very advanced feature which few experts use. We discussed in #19359 removing this access, and it will be deprecated in 5.5.0 (#24691) and removed in 6.0.0 (#19462). The docs have also been replaced with an [example of writing a custom ScriptEngine](https://www.elastic.co/guide/en/elasticsearch/reference/5.x/modules-scripting-engine.html) for this purpose.
